runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -235,31 +251,33 @@ public int appendObjectState(Stream<Map.Entry> entryStream) {
         ImmutableMap<CheckpointEntry.CheckpointDictKey, String> mdkv =
                 ImmutableMap.copyOf(this.mdkv);
 
-        final Iterable<List<Map.Entry>> partitions =
-                Iterables.partition(entryStream::iterator, batchSize);
         int entryCount = 0;
-
-        for (List<Map.Entry> partition : partitions) {
-            MultiSMREntry smrEntries = new MultiSMREntry();
-            for (Map.Entry entry : partition) {
-                smrEntries.addTo(new SMREntry("put",
-                        new Object[]{keyMutator.apply(entry.getKey()),
-                                valueMutator.apply(entry.getValue())},
-                        serializer));
-                entryCount++;
+        int numBytesPerCheckpointEntry = 0;
+        ByteBuf b = Unpooled.buffer();
+        MultiSMREntry smrEntries = new MultiSMREntry();
+
+        for (Map.Entry entry : (Iterable<Map.Entry>) entryStream::iterator) {
+            SMREntry each = new SMREntry("put",
+                    new Object[]{keyMutator.apply(entry.getKey()),
+                            valueMutator.apply(entry.getValue())},
+                    serializer);
+
+            each.serialize(b);
+            numBytesPerCheckpointEntry += b.writerIndex();
+
+            if(numBytesPerCheckpointEntry > rt.getParameters().getMaxWriteSize()) {
```

Comment:
This is not accurate. You need to check for `b.writerIndex() + numBytesPerCheckpointEntry > rt.getParameters().getMaxWriteSize()`, if that is true that you write it out, otherwise you add it to  smrEntries and adjust your accumulator.

Reply:
I would just write out the batch inside the loop body and use `continue`, it simplifies the control-flow logic.

Reply:
+1 to Maithem's suggestion that would ease readability..

Reply:
In order to overlap the resetting of the bookkeeping data structures and utilizing the same loop to consider the SMR entry that is read from the stream (once its read from the stream, the entry cannot be re-read after a "continue" in a loop, I have come up with this control flow. I can add more comments into it for clarification around the control flow.

Reply:
@annym please see my above comment.